lib/_stream_writable.js

```diff
@@ -487,20 +487,20 @@ function onwrite(stream, er) {
       onwriteError(stream, state, er, cb);
     }
   } else {
-    // Check if we're actually ready to finish, but don't emit yet
-    var finished = needFinish(state) || stream.destroyed;
-
-    if (!finished &&
+    if (state.bufferedRequest &&
         !state.corked &&
-        !state.bufferProcessing &&
-        state.bufferedRequest) {
+        !state.bufferProcessing) {
       clearBuffer(stream, state);
     }
 
-    if (sync) {
-      process.nextTick(afterWrite, stream, state, cb);
+    if (state.finished || state.needDrain || cb !== nop) {
```

Comment:
Doesn't do a full needFinish or needDrain check in order to reduce the number of conditions in the hot path.